mne/annotations.py

```diff
@@ -647,6 +647,186 @@ def rename(self, mapping, verbose=None):
+        # create a list of annotations for each epoch
+        epoch_annot_list = [[] for _ in range(len(self.events))]
+
+        # check if annotations exist
+        if self.annotations is None:
+            return epoch_annot_list
+
+        # when each epoch and annotation starts/stops
+        # no need to account for first_samp here...
+        epoch_tzeros = self.events[:, 0] / self._raw_sfreq
+        epoch_starts, epoch_stops = np.atleast_2d(
+            epoch_tzeros) + np.atleast_2d(self.times[[0, -1]]).T
+        # ... because first_samp isn't accounted for here either
+        annot_starts = self._annotations.onset
+        annot_stops = annot_starts + self._annotations.duration
+
+        # get epochs that start within the annotations
+        annot_straddles_epoch_start = np.logical_and(
+            np.atleast_2d(epoch_starts) >= np.atleast_2d(annot_starts).T,
+            np.atleast_2d(epoch_starts) < np.atleast_2d(annot_stops).T)
+
+        # get epochs that end within the annotations
+        annot_straddles_epoch_end = np.logical_and(
+            np.atleast_2d(epoch_stops) > np.atleast_2d(annot_starts).T,
+            np.atleast_2d(epoch_stops) <= np.atleast_2d(annot_stops).T)
+
+        # get epochs that are fully contained within annotations
+        annot_fully_within_epoch = np.logical_and(
+            np.atleast_2d(epoch_starts) <= np.atleast_2d(annot_starts).T,
+            np.atleast_2d(epoch_stops) >= np.atleast_2d(annot_stops).T)
+
+        # combine all cases to get array of shape (n_annotations, n_epochs).
+        # Nonzero entries indicate overlap between the corresponding
+        # annotation (row index) and epoch (column index).
+        all_cases = (annot_straddles_epoch_start +
+                     annot_straddles_epoch_end +
+                     annot_fully_within_epoch)
+
+        # for each Epoch-Annotation overlap occurrence:
+        for annot_ix, epo_ix in zip(*np.nonzero(all_cases)):
+            this_annot = self._annotations[annot_ix]
+            this_tzero = epoch_tzeros[epo_ix]
+            # adjust annotation onset to be relative to epoch tzero...
+            annot = (this_annot['onset'] - this_tzero,
+                     this_annot['duration'],
+                     this_annot['description'])
+            # ...then add it to the correct sublist of `epoch_annot_list`
+            epoch_annot_list[epo_ix].append(annot)
+        return epoch_annot_list
+
+    def add_annotations_to_metadata(self):
+        """Add raw annotations into the Epochs metadata data frame.
+
+        Adds three columns to the ``metadata`` consisting of a list
+        in each row:
+        - ``Annotations_onset``: the onset of each Annotation within
+        the Epoch relative to the start time of the Epoch (in seconds).
+        - ``Annotations_duration``: the duration of each Annotation
+        within the Epoch in seconds.
+        - ``Annotations_description``: the description of each
+        Annotation.
```

Comment:
Sure that works for me. I kept it very verbose because I wonder if there would ever be a case that metadata already has things like `annot_onset`, etc.?

Reply:
if that's a concern, then add something that checks whether those column names already exist, and either (1) error out, or (2) warn that different column names had to be used (like `annot_onset_1` or similar).  To me it seems quite unlikely so I'd be inclined to just error out on the assumption that users will (almost) never hit that corner case.

Reply:
I added this error check and also a `overwrite` parameter to handle the edge case that someone might want to overwrite these columns.